R/utils.R

```diff
@@ -63,10 +63,11 @@ names2 <- function(x) {
 linter_auto_name <- function(which = -3L) {
   call <- sys.call(which = which)
   nm <- paste(deparse(call, 500L), collapse = " ")
-  regex <- rex(start, one_or_more(alnum %or% "." %or% "_"))
+  regex <- rex(start, one_or_more(alnum %or% "." %or% "_" %or% ":"))
   if (re_matches(nm, regex)) {
     match <- re_matches(nm, regex, locations = TRUE)
     nm <- substr(nm, start = 1L, stop = match[1L, "end"])
+    nm <- re_substitutes(nm, rex::rex(start, alnums, "::"), "")
```

Comment:
what happens if user (probably accidentally) runs:

```
lintr::lint(file, linters = list(lintr::assignment_linter, my_custom_package::assignment_linter))
```

where they've written their own `assignment_linter` and perhaps inadvertently included `lintr`' implementation (e.g. picked up from settings?).

i guess that should error up-front, but what's the current behavior?

Reply:
That will work just fine, producing two linters with the same name:
```
> lintr::lint(text = "a = 1", linters = list(assignment_linter(), lintr::assignment_linter()))
<text>:1:3: style: Use <-, not =, for assignment.
a = 1
  ^
<text>:1:3: style: Use <-, not =, for assignment.
a = 1
  ^
```

Having two linters with the same name currently disables partial matching for those, however:
```
> lintr::lint(text = "a = 1 # nolint: a.", linters = list(assignment_linter(), assignment_linter = lintr::assignment_linter()))
Warning in add_exclusions(exclusions, nolints[i], linters_string, exclude_linter_sep,  :
  Could not find linter named ‘a’ in the list of active linters. Make sure the linter is uniquely identified by the given name or prefix.
<text>:1:3: style: Use <-, not =, for assignment.
a = 1 # nolint: a.
  ^
<text>:1:3: style: Use <-, not =, for assignment.
a = 1 # nolint: a.
  ^
> lintr::lint(text = "a = 1 # nolint: assignment_linter.", linters = list(assignment_linter(), assignment_linter = lintr::assignment_linter()))
```

I think there might be a valid use case once #1104 is implemented, namely to have `# nolint: object_name_linter.` suppress multiple versions of the linter, i.e. for different scopes.
WDYT about using `unique(names(linters))` for the partial matching to not disable that for this scenario, but let it happen without an error?
I don't expect end users to be wary of this when they want to configure multiple `object_name_linter()`s

Reply:
> WDYT about using unique(names(linters)) for the partial matching to not disable that for this scenario, but let it happen without an error?

I think the downside is silently ignoring ambiguity in other context. e.g. `# nolint: spaces.` right now matches `spaces_inside_linter` and `spaces_left_parentheses_linter` -- do we want that to work without warning? If that's intentional it seems too clever to me. (same happens with `# nolint: paren`).

So for now, I'd rather keep warning. Maybe once we have an actual use case in hand it'll be clearer what to do.